### ISSUE TITLE
Tests: Fix tests after latest changes to ThaliCore

### ISSUE THREAD

**Issue by mlesnic** [thaliproject/Thali_CordovaPlugin#1924]
Recently, after updates to ThaliCore, when we try to connect to a peer that is not present in NativeLayer availablePeers array, we throw `Peer is unavailable` instead of `Connection could not be established`. In that case, this peer won't be recreated unless we receive a new `peerAvailabilityChanged` event and will add this peer on Native side. 
Because of that we should update `wifiBasedNativeMock` and few tests.
Tested on 2 iOS devices.

**1. multiConnect properly fails on legal but non-existent peerID**
After last changes in ThaliCore, when we try to connect to non existing peer, we throw `Peer is unavailable` instead of `Connection could not be established`. This will cause that we will not recreate this unavailable peer in [thaliMobileNativeWrapper](https://github.com/thaliproject/Thali_CordovaPlugin/blob/master/thali/NextGeneration/thaliMobileNativeWrapper.js#L704) when multiconnect resolves.
`Connection could not be established` is thrown only when we try to connect to available peer and during that, an error occurs.

```# multiConnect properly fails on legal but non-existent peerID
[ThaliCore] BrowserManager.startListeningForAdvertisements
[ThaliCore] Browser.init(serviceType:foundPeer:lostPeer:) peer:4782F6BF-B8DB-42E8-BBCF-E9DF687880A4
[ThaliCore] Browser.startListening
2017-06-28 08:50:41 - DEBUG thaliMobileNativeWrapper: 'discoveryAdvertisingStateUpdateNonTCP: {"discoveryActive":true,"advertisingActive":false}'
2017-06-28 08:50:41 - INFO thaliMobile: 'Received state ({"discoveryActive":true,"advertisingActive":false}) did not match with target ({"started":false,"listening":false,"advertising":false,"networkType":"BOTH","startArguments":{"router":null,"pskIdToSecret":null,"networkType":null}})'
2017-06-28 08:50:41 - INFO thaliMobile: 'Filtered out discoveryAdvertisingStateUpdate (was in stopped state).'
ok 171 No error on starting
ok 172 Got null as expected
[ThaliCore] BrowserManager.connectToPeer(_:syncValue:completion:) peer:67f17d21-d699-4679-89ad-94499596716b
2017-06-28 08:50:41 - DEBUG thaliMobileNativeWrapper: 'multiConnectResolved: {"syncValue":"236cNY6tA4Ni1YCEYFBi101lRg2PdnZj","error":"Peer is unavailable","portNumber":null}'
ok 173 Should only get called after multiConnect returned
ok 174 SyncValue matches
not ok 175 Got right error
  ---
    operator: equal
    expected: 'Connection could not be established'
    actual:   'Peer is unavailable'
  ...
ok 176 listeningPort is null
```

**2. We properly fire peer unavailable and then available when connection fails on iOS**
```
# We properly fire peer unavailable and then available when connection fails on iOS
2017-06-28 09:14:35 - DEBUG thaliMobileNativeWrapper: 'listening 54061'
[ThaliCore] BrowserManager.startListeningForAdvertisements
[ThaliCore] Browser.init(serviceType:foundPeer:lostPeer:) peer:35C3FDE2-BE41-47BD-A427-966F6312A30D
[ThaliCore] Browser.startListening
2017-06-28 09:14:35 - DEBUG thaliMobileNativeWrapper: 'discoveryAdvertisingStateUpdateNonTCP: {"discoveryActive":true,"advertisingActive":false}'
2017-06-28 09:14:35 - DEBUG thaliMobileNativeWrapper: 'Received peer availability changed event with {"peerIdentifier":"fd9211cf-fc7a-4404-9ec8-9ad7df092922","generation":0,"peerAvailable":true}'
2017-06-28 09:14:35 - DEBUG thaliMobileNativeWrapper: 'handlePeerAvailabilityChanged - Emitting {"peerIdentifier":"fd9211cf-fc7a-4404-9ec8-9ad7df092922","peerAvailable":true,"generation":0,"portNumber":null}'
2017-06-28 09:14:35 - DEBUG thaliMobile: 'Emitting peerAvailabilityChanged from handlePeer {"peerIdentifier":"fd9211cf-fc7a-4404-9ec8-9ad7df092922","connectionType":"MPCF","peerAvailable":true,"generation":0,"newAddressPort":false}'
ok 527 peerIds match
ok 528 peer is available
2017-06-28 09:14:35 - DEBUG thaliMobileNativeWrapper: 'Issuing multiConnect for fd9211cf-fc7a-4404-9ec8-9ad7df092922 (syncValue: 0)'
2017-06-28 09:14:35 - DEBUG thaliMobileNativeWrapper: 'Got multiConnect callback'
[ThaliCore] BrowserManager.connectToPeer(_:syncValue:completion:) peer:fd9211cf-fc7a-4404-9ec8-9ad7df092922
2017-06-28 09:14:35 - DEBUG thaliMobileNativeWrapper: 'multiConnectResolved: {"syncValue":"0","error":"Peer is unavailable","portNumber":null}'
2017-06-28 09:14:35 - DEBUG thaliMobileNativeWrapper: 'Received peer availability changed event with {"peerIdentifier":"fd9211cf-fc7a-4404-9ec8-9ad7df092922","peerAvailable":false,"portNumber":null,"recreated":true}'
2017-06-28 09:14:35 - DEBUG thaliMobileNativeWrapper: 'Received peer availability changed event with {"peerIdentifier":"fd9211cf-fc7a-4404-9ec8-9ad7df092922","peerAvailable":true,"generation":0,"portNumber":null,"recreated":true}'
not ok 529 error description matches
  ---
    operator: equal
    expected: 'Connection could not be established'
    actual:   'Peer is unavailable'
  ...
```

**Comment by mlesnic** [comment]
We could also adjust `executeZombieProofTest` to check if error thrown by `connectToPeer` function is equal to `Peer is unavailable` and then if no, just remove it from our array.

**Comment by mlesnic** [comment]
@yaronyg What do you think about changing the error that native layer throws when there is not available peer on the native side? If you agree, would it mean that we should also change this behaviour on Android?

**Comment by mlesnic** [comment]
Also in some cases, we would recreatePeer in an infinite loop right [here](https://github.com/thaliproject/Thali_CordovaPlugin/blob/master/thali/NextGeneration/thaliMobileNativeWrapper.js#L1316), so instead we should check for error and if it is `Peer is unavailable`, we shouldn't recreate it.
So, proposed fix is to modify it as:
```
if (failedConnection.error.message !== 'Peer is unavailable') {
  recreatePeer(failedConnection.peerIdentifier);
}
```

**Comment by yaronyg** [comment]
A few notes:

Infinite Loop - This is intentional. The reasoning is that it's the native layer that decides if a peer is available or not. So if we try to connect to the peer and the connection doesn't work this can be for lots and lots of transient reasons. So we depend on the native layer to distinguish between a transient failure and a 'the peer a'int there no more' failure. So until the native layer says 'the peer is gone' we will keep recreating the peer after connection failures. This is a feature. Not a bug.

Peer Is Not Available - So I don't have time to go through the code right this second but something is wrong here. A connect or multi-connect call should only be able to return 'Connection could not be established'. In fact we don't even have a 'Peer is unavailable' error defined anywhere. And 'connection could not be established' means what it says. That we couldn't connect to the peer. It doesn't mean the peer isn't there.

So personally I think the right thing to do is to keep the issues separate. When we call down to the native layer with connect/multi-connect then we should return 'Connection could not be established'. This could be because the peer really isn't available. Or it could be because of a temporary error. Separately the native layer should be firing up peerAvailabilityChanged messages when the native layer thinks a peer is gone.

So this means we should not ever be returning peer is unavailable from multiconnect on iOS. Only connection could not be established.

Thoughts?

**Comment by enricogior** [comment]
@yaronyg I proposed the add the the `Peer is unavailable` error.

First of all, the assumption that `A connect or multi-connect call should only be able to return 'Connection could not be established'` is incorrect. It's perfectly possible that the native layer receives a request to connect to a peer that is not available when the native layer checks the peer availability. It's a corner case but it's still a valid case and it does actually happen.
Therefore it make sense to have different error messages for different error types.
It adds information for the caller.

If the caller doesn't care about that extra information, it only has to treat the two errors in the same way simply doing `if (error == 'Peer is unavailable' || error == 'Connection could not be established')...`.
But if it does care it can use the extra information to make the code more robust given it can stop trying to connect to an unavailable peer right away instead of relying on some extra mechanism that by the way it seems currently broken: I didn't debug the node layer, but I ran into this problem running tests, so my proposal is actually helping fixing the node layer, I'm not saying the node layer has to fix the problem in this way, but I'm providing a way to do so.

In conclusion, it's always correct to return different error massages for different error types. It's a general approach that should always be used when designing an API. It adds extra information that doesn't prevent the caller to ignore it if it wants. It removes incorrect assumptions that corner cases are not possible.

**Comment by mlesnic** [comment]
Additionally, based on [this](https://github.com/thaliproject/Thali_CordovaPlugin/blob/411576925e169b8546762c74a75c4e750c5d125b/src/ios/AppContext.swift#L362-L380) code, `multiConnectConnectionFailure` is called only when we have any radio off or when we resolve `browserManager.connectToPeer` with error. But in that case `multiConnectResolved` is called too, so we will recreate peer anyway [here](https://github.com/thaliproject/Thali_CordovaPlugin/blob/master/thali/NextGeneration/thaliMobileNativeWrapper.js#L705). In the end, it looks like we could get rid of recreating peer in [multiConnectConnectionFailure](https://github.com/thaliproject/Thali_CordovaPlugin/blob/master/thali/NextGeneration/thaliMobileNativeWrapper.js#L1316) leaving it only in one place, where we check for error.message and if it is equal to `Connection could not be established`, recreate peer.

About throwing only one error by native layer. I think native part should throw various errors depending on given conditions, but we could only react/listen for one error, which is `Connection could not be established`. In that case recreating peer in `multiConnectConnectionFailure` handler without checking for type or cause of error, makes no sense. For example, we could resolve with error about radios turned off. In that case, we don't want to try to reconnect with that peer, we will fail every time and `peerAvailabilityChanged` with this peer's availability set to false, won't be called.

In the end, I think we should leave recreating peer only for `multiConnectResolved` handler.

**Comment by yaronyg** [comment]
@enricogior I didn't mean that the error of 'you asked to connect to peer X and peer X isn't available' doesn't exist. What I meant is that when that happens I believe we should still return 'Connection could not be established'.

The reason for my thinking on this is simplicity. We are an asynchronous system dealing with events that can pop up whenever. We really don't want to make our Node code more complex than it has to be to deal with this. Right now we create a very clear separation of concerns. 

When someone decides to call connect/multi-connect the only thing they should care about is - can this connection be established? If the answer is no, then they need to move on.

Separately they can have an event handler for peerAvailabilityChanged who only concern is 'is this peer available'? So the idea is that if the node layer found out about a peer the only way that could have happened is because a peerAvailabilityChanged event was fired marking the peer as available. So this means that the native layer knows about the peer and is tracking the peer. If the peer goes away then the node layer is allowed to depend on the fact that the native layer will fire another peerAvailabilityChanged specifying that the peer is gone.

This is a powerful simplification at the node layer. For example, in our current code, to find the bug causing the test problem, you should only have to find the peerAvailabilityChanged handler and see why it isn't registering the fact that the peer is gone and thus setting a flag to stop any further connects to that peer.

But if we add a new 'peer is unavailable' error now it's possible for the exact same logic, logic to handle a missing peer, to be in two places in node. One place in the multiconnect call and one place in the peerAvailabilityChanged. And the worst part is that this is all asynchronous code so even if we use the same function to handle the logic in all places that function can be called multiple times in different contexts.

So in the interests of making our code as simple as possible (but no simpler), I think we should keep a clear separation of concerns.

So it seems to me that the real bug isn't that we don't return a peer is unavailable error but rather the real bug is either that the native layer is forgetting to send a peerAvailabilityChanged event OR that the node layer got the event and isn't handling it properly. I think that is where we should focus our debugging efforts.

@mlesnic I think we have some sub-optimal naming but what you are seeing is right. There are two separate things happening here. One thing that is happening is that when we call multiConnect unless we get an immediate error in the callback (which is only reserved for malformed requests, we had to do this in case things like the syncValue aren't defined so we can't correlate the multiConnectResolved event) then we MUST get a multiConnectResolved which indicates what happened to the multiConnect request, including that it failed and how. Now if and only if a connection was actually established and when that connection later is disconnected then we will fire a multiConnectConnectionFailure. This allows us to do things like fire peerAvailabilityChanged events and also allows the peer policy to know that its connection to the peer has gone bad. So the point is that it's correct that both multiConnectResolved and multiConnectConnectionFailure should be fired, but only if a real connection happened. If a real connection didn't happen (e.g. the peer wasn't available as far as the native layer is concerned) then we should only get a multiConnectResolved with 'connection could not be established' but a multiConnectConnectionFailure should never be fired.

This also explains why we can't just put the recreate peer logic into multiConnectResolved. Because:
1. We call multi-connect
2. The multi-connect works just fine and we get a multiConnectResolved with the details
3. The connection later on fails (hence there will be no multiConnectResolved call since it's already fired with the success details)

And now the peer is sorta gone from the node layer.

So we have to have the recreate logic in two places. There really is no way around that.

Except...

While I believe that multi-connect and peerAvailabilityChanged should stay as they are, there is a real question about the recreating peer logic. The whole point of that logic was to make life easier for Thali's users. The 'flow' for them would be 'get notified about peer, put peer in queue, when getting to peer try to connect'. If the connect failed then the peer would be 'recreated' and the whole cycle would start again.

The alternative was a model where we forced the user to manage the peer notifications themselves. That is, we send a peerAvailabilityChanged event and now it's up to the user to decide what to do. If they call connect or multi-connect and it doesn't work then it's up to the user to decide if they want to retry. So long as we don't send another peerAvailabilityChanged event with peerAvailable = false then we are saying we think the peer is still there. This requires more sophisticated management logic from the user (no long just a brain dead simple loop waiting for peer notifications, including fake ones we create as part of the recreate logic) but it makes Thali simpler.

But at this point getting rid of the recreate logic would require a lot of test updates that I'm guessing nobody wants to do.

**Comment by mlesnic** [comment]
@yaronyg Ok, in that case we should leave recreating peer in `multiConnectConnectionFailure`. However, I think we should check also for the type/message of error like it is done in `multiConnectResolved`, because when we receive any error other than `Connection could not be established`, we won't be able to connect anyway.

**Comment by yaronyg** [comment]
@mlesnic Right but even if the error says we can't connect that conclusion should be drawn from a peerAvailabilityChanged event not a multiconnect response.

**Comment by enricogior** [comment]
@yaronyg 

> The reason for my thinking on this is simplicity. We are an asynchronous system dealing with events that can pop up whenever. We really don't want to make our Node code more complex than it has to be to deal with this. Right now we create a very clear separation of concerns.

I agree with that approach. My proposal is actually in that direction.
Given that we also agree on this:

> So the point is that it's correct that both multiConnectResolved and multiConnectConnectionFailure should be fired, but only if a real connection happened. If a real connection didn't happen (e.g. the peer wasn't available as far as the native layer is concerned) then we should only get a multiConnectResolved with 'connection could not be established' but a multiConnectConnectionFailure should never be fired.

There is a bug in the currently code since that is not happening, we actually [fire multiConnectConnectionFailure](https://github.com/thaliproject/Thali_CordovaPlugin/blob/411576925e169b8546762c74a75c4e750c5d125b/src/ios/AppContext.swift#L373-L375) for all error cases, including [ThaliCoreError.startListeningNotActive](https://github.com/thaliproject/thali-ios/blob/77d648fcda9a46142b74b68be85690873047c1a5/ThaliCore/MultipeerConnectivity/BrowserManager.swift#L142) and of course in case of a [peer not found](https://github.com/thaliproject/thali-ios/blob/77d648fcda9a46142b74b68be85690873047c1a5/ThaliCore/MultipeerConnectivity/BrowserManager.swift#L156) in the list of available peers .

So the first fix should be done in [multiConnectToPeer()](https://github.com/thaliproject/Thali_CordovaPlugin/blob/411576925e169b8546762c74a75c4e750c5d125b/src/ios/AppContext.swift#L373-L375) checking the error description and not just if an error occurred. In order to not fire 'multiConnectConnectionFailure' when a peer is not available, BrowserManager.connectToPeer() should return 'Peer is unavailable' and not ['Connection could not be established'](https://github.com/thaliproject/thali-ios/blob/77d648fcda9a46142b74b68be85690873047c1a5/ThaliCore/MultipeerConnectivity/BrowserManager.swift#L154-L158).
Let's for a second assume the the 'Peer is unavailable' will only reach 'multiConnectToPeer()' and not be propagated to the node layer when firing the 'multiConnectResolved' event.
Do we agree on that fix?
If we do I will than explain what we should do with the 'Peer is unavailable' error, since I still thing should be passed to the node layer, but let's skip it for now.

BTW, ThaliCore, [before I fixed it](https://github.com/thaliproject/thali-ios/blob/master-enrico-error-handling-tmp/ThaliCore/MultipeerConnectivity/BrowserManager.swift#L186-L200), never fired 'multiConnectConnectionFailure' if en error occurred during the connection handshaking after the first valid responce from the remote peer (i.e. Session state still equal to 'notConnected' or 'connecting'). Still we don't fire it if the MPC Session object is disconnected after the connection has been successfully established (i.e. Session state equal to 'connected'). There are some valid reasons to not fire 'multiConnectConnectionFailure' after the session state reaches the status of 'connected' but I'm open to reconsider it if you strongly think we should actually fire it.

Moving to the node layer: there is a problem with the current logic of recreatePeer, it's a problem that only occurs in corner cases, but if it occurs it's a showstopper.

1) [recreatePeer()](https://github.com/thaliproject/Thali_CordovaPlugin/blob/master/thali/NextGeneration/thaliMobileNativeWrapper.js#L1121-L1142) breaks the approach we agreed on: it calls 'handlePeerAvailabilityChanged()', but that is supposed to be called by the native layer to notify peer availability changes, so if the node layer also calls it, we can run into race conditions and overwrite what the native layer has communicated to node layer.
2) even if you don't agree with 1), there is also a coding bug since it calls 'handlePeerAvailabilityChanged()' twice in a context where those two calls are supposed to be atomic, so that makes case 1)  even worse since it's wrong to remove and add back a peer without locking the data.

Give 1) and 2) I would like to understand how we can guarantee that the infinite loop will eventually be stopped, because right now I don't see how it can end if we run into a race condition: the native layer notifies the node layer that the peer is gone just before the 'multiConnectConnectionFailure' event is fired, that is not only possible is actually very likely to happen since the connectToPeer call may fail for a timeout error after the MPCF has already received and re-fired a peerLost event.

There is more I would like to discuss, but I guess this is already enough to start with.

**Comment by yaronyg** [comment]
So if we are firing multiConnectConnectionFailure for all error cases then that is a straight up bug as defined [here](https://github.com/thaliproject/Thali_CordovaPlugin/blob/522dfbe7fe4523f63287ea8d5f38e402fa88e334/thali/NextGeneration/thaliMobileNative.js#L521). No connection? No multiConnectConnectionFailure.

And I'm fine with 'Peer is unavailable' only rattling around the iOS layer so long as it becomes a 'Connection could not be established' in the Node layer as you suggested.

If we get back a successful response to a multiConnect request then unless the connection lasts for eternity there MUST be some point where multiConnectConnectionFailure is fired. It doesn't matter why the connection was lost. Only that it was. So the key here isn't the iOS state. The key here is that a multiConnect was sent and a multiConnectResolvedCallback was sent back without an error and with a listeningPort. Once that happens then unless the connection lasts for eternity there MUST be a multiConnectConnectionFailure fired at some point.

Now to the bug. I'm confused. I think there is something wrong with the code. Or more likely I'm just so out of date I don't remember things anymore. But it seems like any time we get a peer.peerAvailable = true we call [this line](https://github.com/thaliproject/Thali_CordovaPlugin/blob/522dfbe7fe4523f63287ea8d5f38e402fa88e334/thali/NextGeneration/thaliMobileNativeWrapper.js#L1048). But shouldn't we only need that line for Android? We shouldn't need it for iOS, right? With iOS we just use the port passed to us by the native layer. We don't need a TCP wrapper. Right?

But it seems like [here](https://github.com/thaliproject/Thali_CordovaPlugin/blob/522dfbe7fe4523f63287ea8d5f38e402fa88e334/thali/NextGeneration/thaliMobileNativeWrapper.js#L1265) we are going to call handlePeerAvailabilityChanged when we get a peerAvailabilityChanged event even from iOS and then we are going to call getPeerPort which will eventually bottom out [here](https://github.com/thaliproject/Thali_CordovaPlugin/blob/522dfbe7fe4523f63287ea8d5f38e402fa88e334/thali/NextGeneration/mux/createPeerListener.js#L346) which is going to do a bunch of Android specific stuff that has nothing to do with iOS.

I'm probably just missing something though, right?

**Comment by enricogior** [comment]
@yaronyg 
> So if we are firing multiConnectConnectionFailure for all error cases then that is a straight up bug as defined here. No connection? No multiConnectConnectionFailure.

OK! We have a common ground now 👍 

Now, regarding:
> Once that happens then unless the connection lasts for eternity there MUST be a multiConnectConnectionFailure fired at some point.

and

> Fires the multiConnectConnectionFailureCallback if a multiConnect connection fails. This failure can include a failure induced by a call to `disconnect`.

I don't understand why we should fire multiConnectConnectionFailure if an explicit disconnect has been called? Why should we try to reconnect to the remote peer even if the application code has explicitly requested to disconnect?
Doing so breaks the ThaliTest tests that explicitly disconnect because it triggers the infinite loop, causing the test to fail with a timeout error in the TestServer.

Maybe the right question is: what is the ultimate goal of the multiConnectConnectionFailure handler?
Because it seems the ultimate goal is to try to reconnect, but it is also used to notify the application code that the disconnect event happened.
But having those two actions together seems wrong.
Why the ThaliMobileNativeWrapper assumes it can decide to try to reconnect when that is ultimately a decision that should be made by the application code? What's the point of notifying the node layer of the disconnect event if ThaliMobileNativeWrapper is also trying to reconnect?
I can't see a scenario where the two things can work well together.
We either notify the application code and let it decide if it want to reconnect, or we try to reconnect without notifying the application code (this second option seems utterly wrong).
So we go back to my previous comment when I asked why the ThaliMobileNativeWrapper calls recreatePeer() since that responsability should be left to the native layer?

Moving on to you last question:

> ... and then we are going to call getPeerPort which will eventually bottom out here which is going to do a bunch of Android specific stuff that has nothing to do with iOS.
> I'm probably just missing something though, right?

On iOS, gServersManager is null, so [getPeerPort()](https://github.com/thaliproject/Thali_CordovaPlugin/blob/522dfbe7fe4523f63287ea8d5f38e402fa88e334/thali/NextGeneration/thaliMobileNativeWrapper.js#L1115-L1118) will always resolve right away calling 'Promise.resolve(null);', and won't call 'gServersManager.createPeerListener()'.
On iOS there is nothing that prevents to enter the infinite loop and nothing that will eventually stop that loop. 
My point is that we should not enter the loop in the first place, ever.
I accept the idea of retrying a few times if an error occurs while handshaking the MC Session connection since I've verified it happens quite frequently that there are transient errors, so it's a good thing to retry there (but just a few times, not forever).
But I don't think ThaliMobileNativeWrapper should ever try to reconnect if the session disconnects after it connected successfully.

**Comment by mlesnic** [comment]
It looks like we enter this infinite loop only during tests. This is because we call `disconnect` at the end of test, which will close active connection with specific peer. However, we might not remove this peer from our `availablePeers`. So the scenario is, when we run test, connect to `peerId` with generation`1`, create relay, then close connection, then run next test, find same peer with same `peerId` but now with generation `0`, we will create connection for this peer with highest generation, which is `1`, because it is still in our `availablePeers` array! In that case, we will fail every time and recreate peer every time in infinite loop, because `peerAvailabilityChanged` event won't be fired, because it is fired only for latest, highest generation.
So we could not call `disconnect` with every peer in teardown, or we could overwrite a callback for `multiConnectConnectionFailure` during the test, or we could somehow clean also `availablePeers` on native side, or we could trigger `lostPeer` handler after `disconnect` call which will eventually trigger `peerAvailabilityChanged` event.

About `getPeerPort()`, I think we are good here, because we will get peer's `portNumber` later, after starting `NotificationAction`.

**Comment by yaronyg** [comment]
Per our phone call today it's worth calling out that the point of the recreate peer is NOT to retry the multi-connect. It's actually there because in Android we had a problem of how we give the user a chance to reconnect since they don't know the port to connect to. Historically connect (as in Android) had to be wrapped in a Mux which had to be created. We didn't have a way to do that from the user side so we had to have the recreate behavior to create a new Mux.

But since the Bluetooth logic is trash anyway if we move to something like M-87 then we will have a multi-connect like experience and the recreate behavior will be completely useless since in multi-connect the port is created dynamically by the multi-connect command. No Mux or other nastiness involved.

So per our conversation we'll do this in steps. Step 1 is to pull the recreate behavior for iOS. We don't need it. Then later we can evaluate Android.
